### DISCUSSION TITLE
404 page re-compiled everytime

### DISCUSSION THREAD

**Discussion by mvkv** [vercel/next.js#44124] [Help] [answered]
I'm working with `NextJS`, I'm using `useContext` and `useEffect` to keep track of login session. I'm noticing that when I navigate from a page to another existing page (ex: `Home` -> `About`), Next doesn't re-compile my pages and only does a client-side routing, so my context is preserved and it's useEffect is not triggered, but when I navigate to a `404 page`, the 404 page is being re-compiled by server everytime, so the state of the context goes briefly on the default state before triggering the useEffect to update it.
Is there a way to avoid re-compiling on 404 page? Like doing it client-side to preserve context?

Looking around the internet I didn't find anything about how the pages and the 404 page are compiled, so: do you have some resources to look at to understand better how NextJS pages and error pages routing and compiling works?

Also, do you know if there's a way to avoid re-compiling the 404 page everytime, to avoid that initial default state?

I reproduced the issue in a [Replit snippet](https://replit.com/@FrancescoMaccar/Test-404?v=1):
Basically I have a ContextProvider in /context.tsx, in the context I have a useEffect that modifies the context state, delayed by 3secs, so you can appreciate my problem).
After the page is loaded and 3secs passed, my context is set to "logged in", if you navigate from "Index" to "About", the context is being preserved, because the routing is client-side, when you navigate to "404 page" (that's simply rendering the same thing as the other two pages), the context is reset and the useEffect is triggered again).

As a solution for avoiding this "state loading" I thought about using SSR to fetch the login status before serving the page, but I don't want to use it, because my aim is to export the static version of the website, so I want to know why 404 routing triggers a full website reloading instead of a client-side routing and if there's a way to avoid it.

**Comment by icyJoseph** [comment]
That codesandox doesn't boot up! I don't want to make any assumptions, so could you please fix it so that I can take a look at the situation? Maybe use `stackblitz` instead, https://stackblitz.com/.

**Reply by mvkv** [reply]
Hello, thank you very much for your reply, I just noticed `codesandbox` has some problems, I also tried `stackblitz` but it seems having some problems with NextJS, so I opted for `replit`, let me know if you can appreciate my issue and if you know some way to avoid it.
I also updated a bit my question, to reflect the new code snippet and with some more useful info on what I want to understand.

**Comment by icyJoseph** [comment, marked as answer]
I see, the problem seems to be that when the router cannot match anything in the page manifest, then it just hands over to a method that does hard navigation, https://github.com/vercel/next.js/blob/canary/packages/next/shared/lib/router/router.ts#L1421:

```ts
    // The build manifest needs to be loaded before auto-static dynamic pages
    // get their query parameters to allow ensuring they can be parsed properly
    // when rewritten to
    let pages: string[], rewrites: any
    try {
      ;[pages, { __rewrites: rewrites }] = await Promise.all([
        this.pageLoader.getPageList(),
        getClientBuildManifest(),
        this.pageLoader.getMiddleware(),
      ])
    } catch (err) {
      // If we fail to resolve the page list or client-build manifest, we must
      // do a server-side transition:
      handleHardNavigation({ url: as, router: this })
      return false
    }
```

I suppose there's some deeper reasoning for this, but I am a bit short on time these days.

I think you can provide a fallback route as a rewrite, so that the state is preserved.

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  async rewrites() {
    return { fallback: [{ source: "/:path*", destination: "/404" }] };
  },
};

module.exports = nextConfig;
```

This might have issues of its own though, but do please try it. The `:path` bit would be available within `useRouter().query` IIRC.


A few other things, using the `getContext` wrapper you have, breaks rules of hooks, which doesn't mean it won't work, but it'll  introduce complications, mostly for developer experience. 

Also, React Components ought to start with a capita letter, aka, `const Index = () => <div/>` not `const index = () => <div/>`

**Reply by mvkv** [reply]
Thank you very much for the reply and for your time! First time going deeper into how NextJS is built. I will try your approach as soon as possible!

UPDATE: I tested it, it works perfectly!
I link the documentation related to rewrites in case someone will need it!

https://nextjs.org/docs/api-reference/next.config.js/rewrites
